chx_compress/io/multifile/multifile.py

```diff
@@ -1,8 +1,304 @@
+            if len(dlen_data) == 0:
+                # there is no image data at this offset
+                #   we have reached the end of this file
+                break
+            else:
+                # there is image data at this offset
+                image_offsets.append(next_image_offset)
+
+                # calculate the offset of the next image (if one exists)
+                dlen = struct.unpack("I", dlen_data)[0]
+                # image data consists of two arrays:
+                #   array of pixel indices
+                #   array of pixel values
+                # we just want to calculate the offset to the start of the next dlen
+                pixel_indices_size = dlen * 4
+                pixel_values_size = dlen * self.header_info["byte_count"]
+
+                # we don't know yet if there will be an image at this offset
+                #   tune in to the next loop iteration to find out!
+                next_image_offset = (
+                    next_image_offset + 4 + pixel_indices_size + pixel_values_size
+                )
+
+        self._image_offsets = tuple(image_offsets)
+
+    def get_image_data(self, image_index):
+        if len(self._image_offsets) == 0:
+            raise ValueError("multifile has no images")
+        elif 0 <= image_index < len(self._image_offsets):
+            self._read_buffer.seek(self._image_offsets[image_index])
+            # read dlen
+            dlen_data = self._read_buffer.read(4)
+            dlen = struct.unpack("I", dlen_data)[0]
+            # what is the size of the corresponding pixel index list and pixel value array
+            pixel_indices_size = dlen * 4
+            pixel_indices_data = self._read_buffer.read(pixel_indices_size)
+            pixel_indices = struct.unpack(f"{dlen}I", pixel_indices_data)
+            pixel_values_size = dlen * self.header_info["byte_count"]
+            pixel_values_data = self._read_buffer.read(pixel_values_size)
+            if self.header_info["byte_count"] == 2:
+                unpack_format = f"{dlen}h"
+            elif self.header_info["byte_count"] == 4:
+                unpack_format = f"{dlen}i"
+            else:
+                unpack_format = f"{dlen}d"
+            pixel_values = struct.unpack(unpack_format, pixel_values_data)
+
+            return pixel_indices, pixel_values
+        else:
+            raise ValueError(
+                f"image_index: {image_index} is not in the range [0, {len(self._image_offsets)})"
+            )
+
+
+def multifile_reader(filepath, mode="rb"):
+    """A convenience function to create a MultifileReader that reads from a file.
+
+    Parameters
+    ----------
+    filepath : str
+        path for the multifile
+    mode : str
+        optional mode to open the file, default is "rb"
+
+    """
+    return MultifileReader(lambda: open(filepath, mode=mode))
+
+
+class MultifileWriter:
+    def __init__(
+        self,
+        write_buffer_factory,
```

Comment:
Why not just a buffer? That would be a common pattern.

In suitcase we use a factory here because we need to support “formats” that span many separate files, but I don’t think that is the code here.